### PR TITLE
Bugfixes

### LINKED ISSUE #8: plugin-parser.js: consider using async
During #5 code review, I agreed that we should use async forEach() when loading plugins. I have not noticed strange slowdowns, actually, but I think it could improve code readability.

### PR DESCRIPTION
**Pull request by 0xPoly**
Closes #8, #18 and #19.

### CODE REVIEW

File: plugin-parser.js [hunk 1]
```diff
@@ -1,37 +1,50 @@
...
+                fs.stat(fullFile, function(err, stats){
+                    if (stats.isDirectory()) {
+                        var mainFile = path.resolve(fullFile) + path.sep + 'main.json';
+                        fs.access(mainFile, fs.R_OK, function(err) {
+                            if (!err) {
+                                plugins.push(loadJson(mainFile));
+                            }
+                            if (plugins.length > 0) {
+                                callback(plugins);
+                            }
```
Comment (bassosimone) [inline plugin-parser.js:37]: The previous code called the callback only once. If I'm reading it correctly, this code invokes the callback many times, every time passing the vector `plugins` with one more plugin added. Why?

Isn't it possible to fire the callback each time we have a new plugin?

File: exec.js [hunk 2]
```diff
@@ -44,15 +48,19 @@ var collectArgs = function(state) {
...
+                    var option = containers[i].getElementsByTagName("select")[0];
+                }
+                console.log("something has gone wrong");
             }
             // to handle radio buttons
-            else if (option.length > 2) {
+            else if (option[0].type == "radio") {
                 for(var k = 0; k < option.length; k++) {
-                    if (option[k].checked && k != 0) {
+                    if (option[k].checked) {
```
Comment (bassosimone) [inline exec.js:59]: Why did you remove `k != 0` from here?

File: exec.js [hunk 3]
```diff
@@ -44,15 +48,19 @@ var collectArgs = function(state) {
...
-            if (option.length == 1 && activation) {
-                var option = containers[i].getElementsByTagName("select")[0];
+            var option = containers[i].getElementsByClassName("arg-buttons")[0]
+                                        .getElementsByTagName("input");
+            // to handle drop down menu
+            if (!option.length) {
+                if (containers[i].getElementsByTagName("select")) {
+                    var option = containers[i].getElementsByTagName("select")[0];
+                }
+                console.log("something has gone wrong");
```
Comment (bassosimone) [inline exec.js]: When running the App, I see this on the console:

```
[7004:0922/233621:INFO:CONSOLE(58)] "something has gone wrong", source: .../network-meter/exec.js (58)
[7004:0922/233625:INFO:CONSOLE(58)] "something has gone wrong", source: .../network-meter/exec.js (58)
[7004:0922/233630:INFO:CONSOLE(58)] "something has gone wrong", source: .../network-meter/exec.js (58)
```

You probably want this to be printed only on `else`. Also, I'd rather throw an exception rather than a warning if `length` is zero and the control is not `select`, so that one notices there is an unexpected error and reports a bug to us.

File: plugin-parser.js (hunk 1 above)
Comment (bassosimone) [inline plugin-parser.js:37]: I've modified the code and can confirm that the callback is called N times for N available plugins, each time with one more plugin added to the vector, starting of course from 1.

**Comment by bassosimone** [comment]
Hi @0xPoly! I finished my review and added comments above.

File: exec.js (hunk 3 above)
Comment (0xPoly) [inline exec.js]: Good catch, fixing.

File: exec.js (hunk 2 above)
Comment (0xPoly) [inline exec.js:59]: Earlier, `option` used to refer to all the inputs for each arg (including the activation checkbox). As such, `k != 0` was needed to make sure we were not triggering on the activation checkbox.

Since the new `options` only contains the argument controls, this is no longer needed.

File: plugin-parser.js (hunk 1 above)
Comment (0xPoly) [inline plugin-parser.js:37]: I think this should fix the extra callbacks. Can you confirm?

File: exec.js (hunk 2 above)
Comment (bassosimone) [inline exec.js:59]: Got it!

File: plugin-parser.js [hunk 4]
```diff
@@ -1,37 +1,50 @@
...
+
+            files.forEach(function(f, index, array) {
+                var fullFile = pluginFolder + path.sep + f;
+                fs.stat(fullFile, function(err, stats){
+                    if (stats.isDirectory()) {
+                        var mainFile = path.resolve(fullFile) + path.sep + 'main.json';
+                        fs.access(mainFile, fs.R_OK, function(err) {
+                            if (!err) {
+                                plugins.push(loadJson(mainFile));
+                                callback(loadJson(mainFile));
```
Comment (bassosimone) [inline plugin-parser.js]: Here I suggest to only load the JSON file once, store the result in a variable, and then reuse it.

Comment (0xPoly) [inline plugin-parser.js]: done!

File: configure.js [hunk 5]
```diff
@@ -33,7 +33,13 @@ exports.init = function() {
...
         pluginParser.loadPlugins(function(data){
             // actions can only be bound to grid after full initialization
-            state.install.set("plugins", data).then(libplugin.bindDeleteButtons(state));
+            var temp_plugins;
+            temp_plugins = state.install.get("plugins");
+            if (temp_plugins == undefined) {
+                temp_plugins = [];
+            }
+            temp_plugins.push(data);
+            state.install.set("plugins", temp_plugins).then(libplugin.bindDeleteButtons(state));
```
Comment (bassosimone) [inline configure.js:42]: Is there a way to bind the delete button of a single plugin, the one passed as argument to data?

Comment (0xPoly) [inline configure.js:42]: Sounds possible. Let me look into it...

Comment (bassosimone) [inline configure.js:42]: To be clear, the sense of my comment is that I'd like to get rid of caching vectors here and just propagate the change specific to that plugin. I'd rather do this because I think it leads to cleaner code.

**Comment by bassosimone** [comment]
Thank you @0xPoly for working on this! Now merging!
